scripts/abins/input/dmol3loader.py

```diff
@@ -41,245 +55,233 @@ def read_vibrational_or_phonon_data(self):
         """
         atoms = {}
-        atom_indx = 0
-        self._parser.find_first(file_obj=file_obj, msg="$coordinates")
-        end_msgs = ["$end", "----------------------------------------------------------------------"]
-
-        while not self._parser.block_end(file_obj=file_obj, msg=end_msgs):
+        atom_index = 0
+        TextParser.find_first(file_obj=file_obj, msg="$coordinates")
+        end_msgs = ["$end", "------------------------------------------------"]
 
-            line = file_obj.readline()
-            entries = line.split()
+        while not TextParser.block_end(file_obj, msg=end_msgs):
+            symbol, *coord = file_obj.readline().split()
 
-            symbol = str(entries[0].decode("utf-8").capitalize())
+            symbol = symbol.decode("utf-8").capitalize()
             atom = Atom(symbol=symbol)
-            # We change unit of atomic displacements from atomic length units to Angstroms
-            au2ang = ATOMIC_LENGTH_2_ANGSTROM
-            float_type = FLOAT_TYPE
 
-            atoms["atom_{}".format(atom_indx)] = {"symbol": symbol, "mass": atom.mass, "sort": atom_indx,
-                                                  "coord": np.asarray(entries[1:]).astype(dtype=float_type) * au2ang}
+            # Convert coordinate units from Bohr to Angstrom
+            coord = np.asarray(coord,
+                               dtype=FLOAT_TYPE) * ATOMIC_LENGTH_2_ANGSTROM
 
-            atom_indx += 1
+            atoms[f"atom_{atom_index}"] = {"symbol": symbol, "mass": atom.mass,
+                                           "sort": atom_index, "coord": coord}
+            atom_index += 1
 
-        self.check_isotopes_substitution(atoms=atoms, masses=masses_from_file)
+        return atoms
 
-        data["atoms"] = atoms
+    @staticmethod
+    def _get_to_next_nonempty(file_obj: io.BufferedReader) -> None:
+        # Scroll to next non-empty line
+        last_line = file_obj.tell()
+        for line in file_obj:
+            if line.decode('utf8').strip():
+                file_obj.seek(last_line)
+                break
+            else:
+                last_line = file_obj.tell()
 
-    def _read_modes(self, file_obj=None, data=None):
+    @classmethod
+    def _read_modes(cls, file_obj) -> Tuple[np.ndarray, np.ndarray]:
+        """Read vibrational frequencies and atomic displacements
+
+        Modes are in their raw format: a 2-D array with modes along index 1
+        (i.e. columns, corresponding to frequencies) and atomic degrees of
+        freedom along index 0 (i.e. rows of x y z x y z x y ... for successive
+        atoms).
         """
-        Reads vibrational modes (frequencies and atomic displacements).
-        :param file_obj: file object from which we read
-        :param data: Python dictionary to which k-point data should be added
+        frequencies, displacements = cls._read_mode_block(file_obj)
+        frequency_blocks = [frequencies]
```

Comment:
could we not initialise the frequency_blocks and displacement_blocks here and then do all of the assignments in the while loop? Then if anything changes in the future its one edit and not 2.

Reply:
Good catch. Can only assume it made more sense in an previous iteration 😅